cmd/compute-domain-controller/node.go

```diff
@@ -0,0 +1,152 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type NodeManager struct {
+	config              *ManagerConfig
+	waitGroup           sync.WaitGroup
+	cancelContext       context.CancelFunc
+	factory             informers.SharedInformerFactory
+	informer            cache.SharedIndexInformer
+	labelCleanupManager *CleanupManager[*corev1.Node]
+}
+
+func NewNodeManager(config *ManagerConfig, getComputeDomain GetComputeDomainFunc) *NodeManager {
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		config.clientsets.Core,
+		informerResyncPeriod,
+		informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+			opts.LabelSelector = metav1.FormatLabelSelector(labelSelector)
+		}),
+	)
+	informer := factory.Core().V1().Nodes().Informer()
+
+	m := &NodeManager{
+		config:   config,
+		factory:  factory,
+		informer: informer,
+	}
+
+	m.labelCleanupManager = NewCleanupManager[*corev1.Node](informer, getComputeDomain, m.cleanupLabels)
+	return m
+}
+
+func (m *NodeManager) Start(ctx context.Context) (rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancelContext = cancel
+
+	defer func() {
+		if rerr != nil {
+			if err := m.Stop(); err != nil {
+				klog.Errorf("error stopping Node manager: %v", err)
+			}
+		}
+	}()
+
+	m.waitGroup.Add(1)
+	go func() {
+		defer m.waitGroup.Done()
+		m.factory.Start(ctx.Done())
+	}()
+
+	if !cache.WaitForCacheSync(ctx.Done(), m.informer.HasSynced) {
+		return fmt.Errorf("NodeManager: informer cache sync failed")
+	}
+
+	if err := m.labelCleanupManager.Start(ctx); err != nil {
+		return fmt.Errorf("NodeManager: error starting labelCleanupManager: %w", err)
+	}
+
+	return nil
+}
+
+func (m *NodeManager) Stop() error {
+	if err := m.labelCleanupManager.Stop(); err != nil {
+		return fmt.Errorf("NodeManager: error stopping labelCleanupManager: %w", err)
+	}
+	m.cancelContext()
+	m.waitGroup.Wait()
+	return nil
+}
+
+// RemoveComputeDomainLabels() searches all nodes that are currently labeled for
+// a specific CD (as identified via CD UID). It then removes that label from all
+// such nodes.
+func (m *NodeManager) RemoveComputeDomainLabels(ctx context.Context, cdUID string) error {
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{cdUID},
+			},
+		},
+	}
+
+	nodes, err := m.config.clientsets.Core.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(labelSelector),
+	})
+	if err != nil {
+		return fmt.Errorf("error retrieving nodes: %w", err)
+	}
+
+	var names []string
+	for _, node := range nodes.Items {
+		if node.Labels[computeDomainLabelKey] != cdUID {
+			continue
+		}
```

Comment:
Hm, now that I look at this again: above, we have this selector:
```
				Key:      computeDomainLabelKey,
				Operator: metav1.LabelSelectorOpIn,
				Values:   []string{cdUID},
```
So -- is it at all possible in here for `node.Labels[computeDomainLabelKey] != cdUID`? I hope this is a noop... :)

Reply:
You're right. I was confusing this with the set of `Nodes` pulled by the informer (which is all `Nodes` having a label of `resource.nvidia.com/computeDomain` for any value).